src/typegoose.providers.ts

```diff
@@ -1,36 +1,79 @@
-import { getModelToken, getConnectionToken } from './typegoose.utils';
-import { TypegooseClass } from './typegoose-class.interface';
-import { Connection, SchemaOptions } from 'mongoose';
-import {getModelForClass} from '@typegoose/typegoose';
+import { FactoryProvider } from '@nestjs/common/interfaces';
+import { getDiscriminatorModelForClass, getModelForClass } from '@typegoose/typegoose';
 import * as isClass from 'is-class';
+import { Connection } from 'mongoose';
+import { TypegooseClass, TypegooseClassWithOptions, TypegooseDiscriminator } from './typegoose-class.interface';
+import { getConnectionToken, getModelToken } from './typegoose.utils';
 
-export type TypegooseClassWithOptions = {
-  typegooseClass: TypegooseClass,
-  schemaOptions?: SchemaOptions
+type ModelFactory = (c: Connection) => any;
+
+export function createTypegooseProviders(connectionName: string,
+                                         models: TypegooseClassWithOptions[] = []): FactoryProvider[] {
+
+  const connectionToken = getConnectionToken(connectionName);
+
+  const buildProvider = ({ name }: TypegooseClass, modelFactory: ModelFactory) => ({
+    provide: getModelToken(name),
+    useFactory: modelFactory,
+    inject: [connectionToken],
+  });
+
+  const createDiscriminatorFactoryFrom = (parentFactory: ModelFactory) =>
+    ({ typegooseClass, discriminatorId }: TypegooseDiscriminator) => buildProvider(
+      typegooseClass,
+      (connection: Connection) => getDiscriminatorModelForClass(
+        parentFactory(connection),
+        typegooseClass,
+        discriminatorId,
+      ),
+    );
+
+  return models.reduce(
+    (providers, { typegooseClass, schemaOptions = {}, discriminators = [] }) => {
+
+      const modelFactory = (connection: Connection) => getModelForClass(
+        typegooseClass,
+        { existingConnection: connection, schemaOptions },
+      );
+
+      const modelProvider = buildProvider(typegooseClass, modelFactory);
+
+      const discriminatorProviders = discriminators.map(createDiscriminatorFactoryFrom(modelFactory));
```

Comment:
`createDiscriminatorFactoryFrom` can only handle a `TypegooseDiscriminator` but here it is provided a union of `TypegooseDiscriminator | TypegooseClass`.

Reply:
Nice catch! Not sure how it was able to compile / pass tests before, though 🤔

Reply:
Yeah, it was really weird when I was updating to master's changes and typescript then decided to error. I did push a commit to fix it though.